community/joystick.py

```diff
@@ -0,0 +1,113 @@
+import pyglet.input
+import math
+import hit
+import time
+import random
+
+class Joystick_controller:
+	def __init__(self, game):
+		self.game = game
+		self.joysticks = pyglet.input.get_joysticks()
+		self.camera_sensitivity = 0.007
+		self.joystick_deadzone = 0.25
+		self.update_delay = 0.1
+		self.last_update = 0
+
+		for joystick in self.joysticks:
+			joystick.on_joybutton_press = self.on_joybutton_press
+			joystick.on_joybutton_release = self.on_joybutton_release
+			joystick.on_joyaxis_motion = self.on_joyaxis_motion
+			joystick.on_joyhat_motion = self.on_joyhat_motion
+			joystick.open()
+
+		self.joystick_pos = [0, 0]
+		self.joystick_rpos = [0, 0]
+		self.joystick_tpos = 0
+		
+		self.joystick_center = [0, 0]
+		self.joystick_rcenter = [0, 0]
+
+		self.joystick_buttons = []
+
+	def update_controller(self):
+		if not self.game.mouse_captured and not self.joysticks:
+			return
+
+		self.game.camera.rotation[0] += self.joystick_rpos[0] * self.camera_sensitivity
+		self.game.camera.rotation[1] += -self.joystick_rpos[1] * self.camera_sensitivity
+
+		self.game.camera.rotation[1] = max(-math.tau / 4, min(math.tau / 4, self.game.camera.rotation[1]))
+
+		if 9 not in self.joystick_buttons and 0 not in self.joystick_buttons:
```

Comment:
How does Pyglet handle button mapping for controllers?

Reply:
Weirdly

Reply:
And I don't think there is a constant for it.

Reply:
Right... I guess I'd have to check it works with other types of controllers (e.g. Xbox and Logitech). You were testing with a PS4 controller, correct?

Reply:
I was testing with a ps4 controller that was being converted to an xbox controller because the ps4's right joystick is centered in a weid way so it always read 1